proxy/coco/src/peer.rs

```diff
@@ -87,60 +77,96 @@ impl Peer {
+            let state = state.lock().await;
             let protocol = state.api.protocol();
             protocol.subscribe().await
         };
         tokio::pin!(protocol_subscriber);
 
         // Start announcement timer.
-        let mut announce_timer = tokio::time::interval(std::time::Duration::from_secs(1));
+        // TODO(xla): Find a more structured approach to manage timings.
+        let mut announce_timer = interval(std::time::Duration::from_secs(1));
+
+        let (announce_sender, mut announcements) =
+            mpsc::channel::<AnnounceEvent>(RECEIVER_CAPACITY);
+        let (peer_sync_sender, mut peer_syncs) = mpsc::channel::<SyncEvent>(RECEIVER_CAPACITY);
+        let (timeout_sender, mut timeouts) = mpsc::channel::<TimeoutEvent>(RECEIVER_CAPACITY);
 
         // Advance the librad protocol.
         tokio::spawn(self.run_loop);
 
+        let mut run_state = RunState::from(run_config);
         loop {
-            let res = tokio::select! {
-                _ = announce_timer.tick() => {
-                    Self::announce(self.state.clone(), &self.store).await.map(Event::Announced)
-                },
-                Some(event) = protocol_subscriber.next() => {
-                    Ok(Event::Protocol(event))
+            let event = tokio::select! {
+                _ = announce_timer.tick() => Event::Announce(AnnounceEvent::Tick),
+                Some(announce_event) = announcements.recv() => Event::Announce(announce_event),
+                Some(protocol_event) = protocol_subscriber.next() => Event::Protocol(protocol_event),
+                Some(sync_event) = peer_syncs.recv() => Event::PeerSync(sync_event),
+                Some(timeout_event) = timeouts.recv() => Event::Timeout(timeout_event),
+                else => {
+                    break
                 },
-                else => break,
             };
 
-            match res {
-                // Propagate if one of the select failed.
-                Err(err) => return Err(err),
-                Ok(event) => {
-                    log::info!("{:?}", event);
-
-                    // Send will error if there are no active receivers. This case is expected and
-                    // should not crash the run loop.
-                    self.subscriber.send(event).ok();
-                },
+            // Send will error if there are no active receivers. This case is expected and
+            // should not terminate the run loop.
+            self.subscriber.send(event.clone()).ok();
+            log::debug!("{:?}", event);
+
+            for cmd in run_state.transition(event) {
+                match cmd {
+                    Command::Announce => {
+                        Self::announce(state.clone(), store.clone(), announce_sender.clone());
+                    },
+                    Command::SyncPeer(peer_id) => {
+                        Self::sync(state.clone(), peer_sync_sender.clone(), peer_id.clone()).await;
```

Comment:
Does this mean we're constantly syncing while the application is live?

Reply:
Nope, this command is only issued when coming online and `sync_on_startup` is set. Also the run_state factors in how many peers to sync with at max and will stop syncing alltogether after a certain period has expired.

Reply:
Nice :ok_hand: So if I'm understanding correctly, that startup logic is isolated in the `run_state` module?

Reply:
That's correct, this loop just advances the `run_state` and executes any commands if any.

Reply:
I dig it :+1: